### PR TITLE
Update config fails to migrate from 1.2.1 to 1.3.0 - Closes #2542

### LINKED ISSUE #2542: Update config fails to migrate from 1.2.1 to 1.3.0
### Expected behavior
Node upgrade should migrate the `config` consistently for every release.
Having some unit test case would be nice way to ensure if everything works before we do the QA process, this way we can ensure the quality at development stage itself.
possible test scenarios:
```
Migrate from testnet versions
1.0.0-rc.x  to  1.x.x-rc.x
Mainnet
1.0.0 to 1.x.x
```
### Actual behavior
Currently, it doesn't migrate the config properly, it adds invalid config data after migration.

```
bash installLisk.sh upgrade -r main -0 no -d /home/lisk -f lisk-1.3.0-alpha.3-a637899-Linux-x86_64.tar.gz
Using local binary lisk-1.3.0-alpha.3-a637899-Linux-x86_64.tar.gz
Stopping Lisk to perform a backup
Lisk configured for mainnet
[+] Lisk stopped successfully.
[+] Postgresql stopped successfully.
Lisk configured for mainnet
[PM2][WARN] No process found
┌──────────┬────┬──────┬─────┬────────┬─────────┬────────┬─────┬─────┬──────┬──────────┐
│ App name │ id │ mode │ pid │ status │ restart │ uptime │ cpu │ mem │ user │ watching │
└──────────┴────┴──────┴─────┴────────┴─────────┴────────┴─────┴─────┴──────┴──────────┘
 Use `pm2 show <id|name>` to get more details about an app
[PM2] [-] Stopping Modules
[PM2] [v] Modules Stopped
[PM2] [-] Stopping all Applications
[PM2][WARN] No process found
[PM2] [v] All Applications Stopped
[PM2] [-] Stopping Agent
[PM2] [v] Agent Stopped
[PM2] [-] Stopping PM2 Daemon
[PM2] [-] PM2 Daemon Stopped
Backing up existing Lisk directory
Extracting Lisk binaries to /home/lisk/lisk-main
Restoring Database to new Lisk Install
Copying config.json entries from previous installation
Applying migration of lisk config file from 1.2.1 to 1.3.0


- No changes for version "undefined"


- No changes for version "0.9.x"


- Applying changes for version "1.0.0-rc.1"
--- renamed port to httpPort
--- added new config wsPort
--- updated db.poolSize to min and max
--- removed peers.options.limits
--- renamed transactions.maxTxsPerQueue to transactions.maxTransactionsPerQueue
--- removed loading.verifyOnLoading
--- removed dapp
--- updated port to wsPort for peers.list
--- added db.logFileName for db related logs
--- added api.options.cors to manage CORS settings for http API
--- added broadcasts.active=true as boolean value to enable/disable the broadcasting behavior
--- convert forging.secret to forging.delegates
No forging secret. So skipping conversion.


- Applying changes for version "1.0.0-rc.2"
--- moved ssl to api.ssl to make SSL as API only config


- No changes for version "1.0.0-rc.3"


- No changes for version "1.0.0-rc.4"


- No changes for version "1.0.0-rc.5"


- No changes for version "1.0.0"


- Applying changes for version "1.1.0-rc.0"
--- removed version and minVersion
--- removed nethash


- No changes for version "1.1.0"


- No changes for version "1.1.1-rc.x"


- No changes for version "1.1.1"


- Applying changes for version "1.2.0-rc.x"
--- remove malformed peer list items


- No changes for version "1.2.0"


- No changes for version "1.2.1"


- No changes for version "1.2.1-rc.x"

Writing updated configuration to /home/lisk/lisk-main/config.json
Starting Lisk with current blockchain
Lisk configured for mainnet
[+] Postgresql started successfully.
[+] Lisk started successfully.
[+] Lisk is running as PID: 23591
Current Block Height:  7548837
cat lisk-main/config.json
{
	"wsPort": null,
	"db": {
		"logFileName": "logs/lisk.log"
	},
	"api": {},
	"peers": {
		"list": []
	},
	"transactions": {}
```
### Steps to reproduce
Upgrade mainnet node from `1.2.1` to `1.3.0`
### Which version(s) does this affect? (Environment, OS, etc...)
`>= 1.3.0`

### PR DESCRIPTION
**Pull request by nazarhussain**
### What was the problem?

It was required to register every version with in the migration script before releasing. 

### How did I fix it?

Removed the dependency of version registration.  
- If no version specified now, it will migrate from beginning till end. 
- If start version provided and not available it will return same configuration 

### How to test it?
```
echo "{}" > config.json
node scripts/update_config.js ./config.json -n testnet -o config.json 1.2.1 1.3.0-alpha.3
```

### Review checklist

* The PR resolves #2542
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: helpers/json_history.js [hunk 1]
```diff
@@ -103,69 +103,93 @@ function JSONHistory(title, logger) {
...
-		if (toVersion === undefined && cb === undefined) {
-			cb = fromVersion;
-			includeStart = true;
-			fromVersion = versions[0];
-			toVersion = versions[versions.length - 1];
+	/**
+	 * Migrate the json object provided
+	 *
+	 * @param {Object} json - JSON object to apply migrations
+	 * @param {string} [jsonVersion] - Start version
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply helpers/json_history.js]: To me it make more sense now, that you specify the version of the json you provided.

**Comment by nazarhussain** [comment]
@ManuGowda If no applicable migration found, it simply write back the original json file. It is necessary to write the output even if nothing changes.

File: helpers/json_history.js [hunk 2]
```diff
@@ -103,69 +103,93 @@ function JSONHistory(title, logger) {
...
+		if (startIndex === undefined) {
+			startIndex = versions.findIndex(version =>
+				semver.ltr(jsonVersion, version)
 			);
 		}
 
-		startFromVersion += includeStart ? 0 : 1;
+		if (startIndex < 0) {
+			self.logger.info(
+				`No migration found applicable from version "${jsonVersion}"`
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply helpers/json_history.js]: If user didn't provided the ending version, then this message may confuse him.

**Comment by ManuGowda** [comment]
When I upgrade a node which has custom config, it copies the custom config, however the logs state different
```
Applying migration of lisk config file from 1.3.0-rc.0 to 1.3.0-rc.1
No migration found applicable from version "1.3.0-rc.0"

Writing updated configuration to /home/lisk/lisk-test/config.json
```

Here is the detailed logs
```
lisk@ams3-manu-testnet-01:~$ bash installLisk.sh upgrade -0 no -r test -d /home/lisk -f lisk-1.3.0-alpha.3-a637899-Linux-x86_64.tar.gz
Using local binary lisk-1.3.0-alpha.3-a637899-Linux-x86_64.tar.gz
Stopping Lisk to perform a backup
Lisk configured for testnet
[+] Lisk stopped successfully.
[+] Postgresql is not running.
Lisk configured for testnet
[PM2][WARN] No process found
┌──────────┬────┬──────┬─────┬────────┬─────────┬────────┬─────┬─────┬──────┬──────────┐
│ App name │ id │ mode │ pid │ status │ restart │ uptime │ cpu │ mem │ user │ watching │
└──────────┴────┴──────┴─────┴────────┴─────────┴────────┴─────┴─────┴──────┴──────────┘
 Use `pm2 show <id|name>` to get more details about an app
[PM2] [-] Stopping Modules
[PM2] [v] Modules Stopped
[PM2] [-] Stopping all Applications
[PM2][WARN] No process found
[PM2] [v] All Applications Stopped
[PM2] [-] Stopping Agent
[PM2] [v] Agent Stopped
[PM2] [-] Stopping PM2 Daemon
[PM2] [-] PM2 Daemon Stopped
Backing up existing Lisk directory
Extracting Lisk binaries to /home/lisk/lisk-test
Restoring Database to new Lisk Install
Copying config.json entries from previous installation
Applying migration of lisk config file from 1.3.0-rc.0 to 1.3.0-rc.1
No migration found applicable from version "1.3.0-rc.0"

Writing updated configuration to /home/lisk/lisk-test/config.json
Starting Lisk with current blockchain
Lisk configured for testnet
[+] Postgresql started successfully.
[+] Lisk started successfully.
[+] Lisk is running as PID: 3166
Current Block Height:  6713512
lisk@ams3-manu-testnet-01:~$ cat lisk-test/config.json
{
	"manu": "gowda",
	"aa": 1234
}lisk@ams3-manu-testnet-01:~$ cat lisk-test/package.json
{
	"name": "lisk",
	"version": "1.3.0-rc.1",
```

File: helpers/json_history.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ManuGowda) [reply helpers/json_history.js]: ok, so they why are we not using the same param names here? we can keep both caller and calee consistant?
https://github.com/LiskHQ/lisk/blob/6ad6cd98c11d688a2cb451f48ec3843ea94f1883/scripts/update_config.js#L265-L269

**Comment by nazarhussain** [comment]
@ManuGowda Updated the message as discussed.

**Comment by nazarhussain** [comment]
@MaciejBaj Added the pre-validation for config file before writing it.

**Comment by MaciejBaj** [comment]
> @nazarhussain if I run `node scripts/update_config.js -n testnet ./config.json 1.2.0-rc.1 1.3.0-rc.0`
> I am getting following results
> 
> ```
> node scripts/update_config.js -n testnet ./config.json 1.2.0-rc.1 1.3.0-rc.0
> Applying migration of lisk config file from 1.2.0-rc.1 to 1.3.0-rc.0
> 
> 
> - Applying changes for version "1.2.0-rc.x"
> --- remove malformed peer list items
> 
> Writing configuration file temporarily for validation.
> /private/var/folders/rf/127w3hbj2rd042bf3dg28gvr0000gn/T/5eda1d57a57400f040fa09a776d22e0b
> 
> Validating the configuration file
> Validation finished successfully.
> 
> 
> ------------ OUTPUT -------------
> {
> 	"wsPort": 65535,
> 	"httpPort": 65535
> }
> ```
> why it doesn't show me the message `Writing configuration file to`?

@ManuGowda without the `output` specified the result will only be printed to console, so it's all good.

**Comment by MaciejBaj** [comment]
@nazarhussain can we do it without introducing the new dependency - `tempy` just for the purpose of storing a temporary config file in update_script?

**Pull request merged**